model/histogram/float_histogram.go

```diff
@@ -438,6 +440,15 @@ func (h *FloatHistogram) Compact(maxEmptyBuckets int) *FloatHistogram {
 // information can be read directly from there rather than be detected each time
 // again.
 func (h *FloatHistogram) DetectReset(previous *FloatHistogram) bool {
+	if h.CounterResetHint == CounterReset {
+		return true
+	}
+	if h.CounterResetHint == NotCounterReset {
+		return false
+	}
+	// In all other cases of CounterResetHint, we go on as we would otherwise.
+	// Even in the GaugeHistogram case, we pretend this is a counter histogram
+	// for consistency.
```

Comment:
How can the hint be something other than the earlier two options

Reply:
Yeah, the explanation should be more comprehensive here. We just did not have enough time during the group coding. Here is my suggestion:

"If the CounterResetHint is UnknownCounterReset, we do not know yet if this histogram comes with a counter reset. Therefore, we have to do all the detailed work to find out if there is a counter reset or not. We do the same if the CounterResetHint is GaugeType. The latter should not happen, but PromQL still allows the user to apply functions to gauge histograms that are only meant for counter histograms. In this case, we treat the gauge histograms as a counter histograms (and we plan to return a warning about it to the user)."

We should probably also adjust the doc comment above to reflect that the DetectReset method will shortcut to true or false if the CounterResetHint is CounterReset or NotCounterReset, respectively. Plus, the last paragraph of the doc comment can be removed, because this PR does what's requested there.